### PR TITLE
Improve check functions

### PR DESCRIPTION
**Pull request by yasahi-hpc**
Improves #80

This PR aims at extending some check functions to work on `std::array` or `std::vector`. A check function `are_valid_axes` has been introduced. For this purpose, a new trait to extract the base value type has been added.

Following modifications are made
- [x] Add a trait `base_container_value_type` to extract the base value type of a container
- [x] Update existing check functions to work on `std::array` or `std::vector`.
- [x] Introduce a new check function `are_valid_axes` to check if the axes are valid for input Views

### CODE REVIEW

File: common/src/KokkosFFT_utils.hpp
Comment (cedricchevalier19) [inline common/src/KokkosFFT_utils.hpp:107]: Not related but for this function you can do an early exit in the for loop.

File: common/src/KokkosFFT_utils.hpp
Comment (cedricchevalier19) [inline common/src/KokkosFFT_utils.hpp:131]: Add a comment on why `int` is good datatype choice.

File: common/src/KokkosFFT_utils.hpp
Comment (cedricchevalier19) [inline common/src/KokkosFFT_utils.hpp:132]: I do not think this `std::vector` is necessary. Its length is known at compile time. You should use an array or something like this if you need "temporary" storage.

File: common/src/KokkosFFT_utils.hpp
Comment (cedricchevalier19) [inline common/src/KokkosFFT_utils.hpp:139]: You can check for `out of range` in this loop directly.

File: common/src/KokkosFFT_utils.hpp
Comment (cedricchevalier19) [inline common/src/KokkosFFT_utils.hpp:147]: So the `non_negative_axes` is used only to check for duplication?

File: common/src/KokkosFFT_utils.hpp
Comment (cedricchevalier19) [inline common/src/KokkosFFT_utils.hpp:142]: What kind of exception do you expect? It might dangerous to hide all the exceptions.

File: common/src/KokkosFFT_utils.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp:107]: That is true. Fixed accordingly

File: common/src/KokkosFFT_utils.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp:134]: In the current implementation, axes type is an alias to `std::array<int, DIM>`. Thus, we use `int` here. I have added a comment

File: common/src/KokkosFFT_utils.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp]: Actually, `out of range` check has already made in `convert_negative_axis` function, which throws a `runtime_error` for `out of range`. Thus, if there is no exception, it is ensured that the axes are not `out of range`. I have added a comment

File: common/src/KokkosFFT_utils.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp]: As you said, we only need duplication check here. Fixed accordingly.

File: common/src/KokkosFFT_utils.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp]: We may have `std::runtime_error` here. Fixed to capture `std::runtime_error` here. Also comments to explain the expected behaviors are added

**Comment by yasahi-hpc** [comment]
@cedricchevalier19 Thanks for reviews. I have fixed accordingly. Can I have a second review please

File: common/src/KokkosFFT_utils.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp:142]: Yes. We are still storing the non-negative axes which are converted from the original axes. Maybe better to rename this method `convert_negative_axis`, which returns positive `axis` as is and negative axis to be non-negative axis.

Let me explain the definition. `DIM` is dimension of FFT (1D to 3D).
The value `axes` is defined by the `std::array<int, DIM>` to specify the axes to perform FFT. 
Each component of axes can take a range from `[-rank, rank-1] ` where `rank` is a `rank` of `View`.

For example, for 3D and 4D Views we may have the following axes

```C++
axes = {0, 2, -2} // Converted to {0, 2, 1} for 3D View
axes = {0, 2, -1} // Converted to {0, 2, 3} for 4D View
```

Comment (cedricchevalier19) [reply common/src/KokkosFFT_utils.hpp:142]: I do not quite follow the logic here.
```C++
axes = {0, 2, -2}; // {0, 2, 3-2=1} in 3D
axes = {0, 2, 1}; // Same than above?
```
However, I do not think it is related to this PR, I simply have to read the documentation.

**Review by cedricchevalier19 — APPROVED** [review]

File: common/src/KokkosFFT_utils.hpp
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp:142]: The logic is the same as `numpy.ftt` but I have not documented this well. I will improve the docs to describe this point.

**Pull request merged**
